### PR TITLE
Backport NuttX register based stack overflow protection 

### PR DESCRIPTION
**Pull request by thapav**
This patch >
1) Removes existing register based stack protection.
2) Adds NuttX based stack protection.

### CODE REVIEW

**Comment by sunghan-chang** [comment]
@thapav Could you share test result with new code? How about PR #4587 ?

**Comment by sangam-swami** [comment]
> @thapav Could you share test result with new code? How about PR #4587 ?

PR #4587 will be verified and updated after this PR.

**Comment by thapav** [comment]
The test results with PR 4587 are as follows;

#Test Stack Protection for Kernel Thread
1) Test Stack access within range: Test stack access within range using array size - Passed !!
2) Test Stack access in overflow: 
  i)  Test Overflow using array declaration: Board reboots
  ii) Test Overflow using recursive function: Board does not reboot

#Test Stack Protection for User Thread [Test Stack protection with CPU register]
1) Test Stack access within range: Test stack access within range using array size - Passed !!
2) Test Stack access in overflow: 
  i)  Test Overflow using array declaration: Board reboots
  ii) Test Overflow using recursive function: Board does not reboot

**Comment by sunghan-chang** [comment]
> ii) Test Overflow using recursive function: Board does not reboot

How about above with mpu based protection? Is it a sample problem or a problem of register based protection?

**Comment by thapav** [comment]
> How about above with mpu based protection? Is it a sample problem or a problem of register based protection?

MPU based reboots fine.

**Comment by sunghan-chang** [comment]
> MPU based reboots fine.

It means that reg based stack protection does not work well. Right? Could you search arm docs to check whether the register supports or not?

**Comment by thapav** [comment]
> > MPU based reboots fine.
> 
> It means that reg based stack protection does not work well. Right? Could you search arm docs to check whether the register supports or not?

Based on arm documentation, it says that:
Stack limit registers are available for all Stack Pointers in ARMv8-M architecture with Main Extension. In ARMv8-M architecture the stack limit registers are available for Secure Stack Pointers only.
To separate Secure and Non-secure stacks, processors that are based on the ARM v8-M architecture support four stack pointers if the ARMv8-M architecture with Security Extension is implemented.

Link:
https://developer.arm.com/documentation/100688/0101/ARMv8-M-architecture-technical-overview/Stack-and-stack-limit-registers

**Comment by thapav** [comment]
Also as per this Armv8-m technical overview document (Pg 11, Para 1):
https://community.arm.com/cfs-file/__key/telligent-evolution-components-attachments/01-2142-00-00-00-00-66-90/Whitepaper-_2D00_-ARMv8_2D00_M-Architecture-Technical-Overview.pdf

*For ARMv8-M Mainline, all stack pointers have corresponding stack limit registers. For ARMv8-M Baseline, Secure stack pointers have corresponding stack limit registers; Non-Secure programs can use the MPU for stack overflow prevention.*

File: os/arch/arm/src/armv8-m/up_svcall.c [hunk 1]
```diff
@@ -273,10 +273,8 @@ int up_svcall(int irq, FAR void *context, FAR void *arg)
 				up_mpu_set_register(&rtcb->mpu_regs[i]);
 			}
 		}
-#if defined(CONFIG_MPU_STACK_OVERFLOW_PROTECTION)
-		up_mpu_set_register(rtcb->stack_mpu_regs);
-#elif defined(CONFIG_REG_STACK_OVERFLOW_PROTECTION)
-		set_PSPLIM((uint32_t) rtcb->stack_alloc_ptr);
+#ifdef CONFIG_MPU_STACK_OVERFLOW_PROTECTION
+		up_mpu_set_register(tcb->stack_mpu_regs);
```
Comment (amandeepc) [inline os/arch/arm/src/armv8-m/up_svcall.c:277]: ** rtcb->stack_mpu_regs

Comment (amandeepc) [reply os/arch/arm/src/armv8-m/up_svcall.c:277]: without this, I get build error

Comment (thapav) [reply os/arch/arm/src/armv8-m/up_svcall.c]: Thanks, Updated it.

**Comment by sunghan-chang** [comment]
@thapav @sangam-swami So, what's your plan for this?

**Comment by sangam-swami** [comment]
> @thapav @sangam-swami So, what's your plan for this?

Realtek project has armv8-m mainline support. In that case "For ARMv8-M Mainline, all stack pointers have corresponding stack limit registers", so it means there are 4 registers available for secure and non-secure stack pointers.
The current method of implementation is correct,  so we will debug the issue on recursive case device not rebooting because of memory corruption

**Comment by thapav** [comment]
> > @thapav @sangam-swami So, what's your plan for this?
> 
> Realtek project has armv8-m mainline support. In that case "For ARMv8-M Mainline, all stack pointers have corresponding stack limit registers", so it means there are 4 registers available for secure and non-secure stack pointers.
> The current method of implementation is correct, so we will debug the issue on recursive case device not rebooting because of memory corruption

Yes I'm currently checking that issue

File: os/arch/arm/src/armv8-m/up_initialstate.c [hunk 2]
```diff
@@ -114,6 +114,15 @@ void up_initial_state(struct tcb_s *tcb)
 	/* Set the stack limit value */
 
 	xcp->regs[REG_R10] = (uint32_t)tcb->stack_alloc_ptr + 64;
+#endif
+#ifdef CONFIG_REG_STACK_OVERFLOW_PROTECTION
+	xcp->regs[REG_SPLIM]   = (uint32_t)tcb->stack_alloc_ptr;
+#endif
+#ifdef CONFIG_ARMV8M_TRUSTZONE
```
Comment (kishore-sn) [inline os/arch/arm/src/armv8-m/up_initialstate.c:121]: This is related to Trustzone and has been removed in a different PR which we raised for fixing TZ context issues. 
Please remove this change as its not related to SP LIM register.

Comment (thapav) [reply os/arch/arm/src/armv8-m/up_initialstate.c]: modified

**Comment by thapav** [comment]
The latest test results with PR 4587 are as follows;

1) MPU based stack protection: Works well for all TC's
2) Register based stack protection:
Test Stack access within range: Test stack access within range using array size - Passed !!
Test Stack access in overflow:
i) Test Overflow using array declaration: Board reboots
ii) Test Overflow using recursive function: Board does not reboot

**Comment by sunghan-chang** [comment]
@thapav Could you add that in commit description?

1. MPU based stack protection: Works well for all TC's in apps/examples/xxxx
2. Register based stack protection:
Test Stack access within range: Test stack access within range using array size - Passed !!
Test Stack access in overflow:
i) Test Overflow using array declaration: Board reboots
ii) Test Overflow using recursive function: Board does not reboot

**Comment by thapav** [comment]
> 1. Is it the same for kernel and user?
Yes the results for kernel TC are the same.

**Comment by sunghan-chang** [comment]
@thapav There are some changes on exception handler. Could you verify the kernel testcase with / without FPU?

**Comment by thapav** [comment]
> @thapav There are some changes on exception handler. Could you verify the kernel testcase with / without FPU?

I checked with latest code , with/without FPU 
Results are the same

**Review by sunghan-chang — APPROVED** [review]

**Pull request merged**